stellar-contract-env-host/src/host.rs

```diff
@@ -794,18 +826,43 @@ impl Host {
                 ))
             }
         };
-        let vm = Vm::new(self, id, code.as_slice())?;
+        let vm = Vm::new(self, id.clone(), code.as_slice())?;
         // Resolve the function symbol and invoke contract call
         vm.invoke_function_raw(self, SymbolStr::from(func).as_ref(), args)
     }
 
+    fn try_call_vm(
+        &self,
+        id: &Hash,
+        func: &Symbol,
+        args: &[RawVal],
+    ) -> Option<Result<RawVal, HostError>> {
+        #[cfg(feature = "vm")]
+        return Some(self.call_vm(id, func, args));
+        #[cfg(not(feature = "vm"))]
+        return None;
+    }
+
+    fn call_n(&self, contract: Object, func: Symbol, args: &[RawVal]) -> Result<RawVal, HostError> {
+        // Get contract ID
+        let id = self.visit_obj(contract, |bin: &Vec<u8>| {
+            let arr: [u8; 32] = bin.as_slice().try_into().map_err(|_| {
+                HostError::WithStatus(
+                    String::from("invalid contract hash"),
+                    ScStatus::HostObjectError(ScHostObjErrorCode::ContractHashWrongLength),
+                )
+            })?;
+            Ok(xdr::Hash(arr))
+        })?;
+
+        self.try_call_vtable(&id, &func, args)
+            .or_else(|| self.try_call_vm(&id, &func, args))
+            .unwrap_or_else(|| unimplemented!())
```

Comment:
Nitpick: I don't feel the vm call should be "nested" inside the vtable. They are more like parallel relationship instead of one being the fallback of the other (which `or_else` implies). 
If they are nested, then if feels more appropriate the other way around (`try_call_vm.or_else(try_call_vtable)`)? Since VM is the more default/production focused branch.

Reply:
I think once we have native contracts the relationship might exist as such where we execute a vtable loaded contract instead of a vm one if it is present. In tests we are really just presenting the test contract as a native contract.

Reply:
I wouldn't call this nested, it's sequential. That is the appropriate relationship because I can only call one function so there must exist an order in which they are considered. This `or_else` is just a very compact expression of this relationship. As @leighmcculloch points out, if native functions were here then the order would be "test, native, vm".

But I disagree that it should be the other way around. If I've registered a test contract, I expect it to get called so I can debug appropriately.

Reply:
For the record, I removed this code but the new behavior is semantically very similar. The only difference is that in the final case it returns a host error instead of panic with `unimplemented`.